test/common/upstream/cluster_manager_impl_test.cc

```diff
@@ -838,4 +838,33 @@ TEST(ClusterManagerInitHelper, AddSecondaryAfterSecondaryInit) {
   cluster2.initialize_callback_();
 }
 
+TEST(ClusterManagerInitHelper, RemoveClusterWithinInitLoop) {
+  InSequence s;
+  ClusterManagerInitHelper init_helper;
+
+  // Tests the scenario encountered in Issue 903: The cluster was removed from
+  // the secondary init list while traversing the list.
+  NiceMock<MockCluster> cluster;
+  // Fake that we are in the secondary InitializePhase for this test.
+  ON_CALL(cluster, initializePhase()).WillByDefault(Return(Cluster::InitializePhase::Secondary));
+  // We expect that initialize() should be called first upon addCluster.  In the
```

Comment:
The code change looks good (thanks for fixing), but I think the test and the explanation can be simplified a bit. For secondary clusters, initialize() is not called when addCluster() is called. It's called once all the primary clusters have initialized OR when static load completes if there are no primary clusters. So, I would rejigger this test to do:

```
ON_CALL(cluster, initializePhase())...
init_helper.addCluster(cluster);
EXPECT_CALL(cluster, initialize())... // remove inline
init_helper.onStaticLoadComplete();
```

Then also simplify the explanation surrounding the test.